scripts/ci/check_license.py

```diff
@@ -0,0 +1,315 @@
+        Check if a license is allowed for a file. The file is relative to the west workspace.
+        The license identifier can be prefixed with a '-' to check if the license is allowed,
+        but with a warning.
+        '''
+        file_path = str(file_path).replace('\\', '/')
+        license_identifier = license_identifier.upper()
+        if license_identifier not in self.allow_list:
+            return False
+        allow = False
+        for pattern, value in self.allow_list[license_identifier]:
+            if pattern.search(file_path):
+                allow = value
+        return allow
+
+class PatchLicenseChecker:
+    '''Check licenses for a git patch.'''
+
+    args: dict
+    license_checker: FileLicenseChecker
+    git_top: Path
+    west_workspace: Path
+    junit: str
+    total_tests: int
+    total_skipped: int
+    total_errors: int
+    total_warnings: int
+
+    def __init__(self, args: dict):
+        self.args = args
+        self.license_checker = FileLicenseChecker(args.allow_list)
+
+    def run(self, program: str, *args: 'list[str|Path]', cwd=None) -> str:
+        '''A helper function to run an external program.'''
+        run_cmd = (program,) + tuple(str(a) for a in args)
+        run_str = ' '.join(shlex.quote(arg) for arg in run_cmd)
+        try:
+            process = subprocess.Popen(run_cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE,
+                                       cwd=cwd)
+        except OSError as e:
+            self.report('error', f'Failed to run "{run_str}": {e}')
+            sys.exit(2)
+        stdout, stderr = process.communicate()
+        stdout = stdout.decode('utf-8')
+        stderr = stderr.decode('utf-8')
+        if process.returncode or stderr:
+            self.report('error', f'Command "{run_str}" exited with {process.returncode}\n'
+                                 f'==stdout==\n{stdout}\n==stderr==\n{stderr}')
+            sys.exit(2)
+        return stdout.rstrip()
+
+    def report(self, label: str, message: str, file_name:'str|Path' = '<none>', license:str = ''):
+        '''Report results to the user.'''
+        file_name = str(file_name)
+        # Print to stdout with optional GitHub Actions Workflow commands.
+        if label not in ('error', 'warning'):
+            print(f'{label.upper()}: {file_name}: {message}')
+        else:
+            if self.args.github:
+                file_path = (self.west_workspace / file_name).relative_to(self.git_top)
+                print(f'::{label} file={file_path},title=License Problem::{file_path}: ' +
+                      message.replace('%', '%25').replace('\r', '%0D').replace('\n', '%0A'))
+            else:
+                print(f'{label.upper()}: {file_name}: {message}', file=sys.stderr)
+        # Put result in JUnit file.
+        license_xml = f':{html.escape(license)}' if license else ''
+        if label == 'error':
+            self.junit += (f'<testcase name="{html.escape(file_name)}{license_xml}" '
```

Comment:
would using 'junit-xml' make this simpler?

Reply:
`junit-xml` looks good, I will try it.

Reply:
done